### PR TITLE
DV2 TyperDeduper: Extract migrations to separate method

### PR DESCRIPTION
**Pull request by edgao**
Based on top of https://github.com/airbytehq/airbyte/pull/35342

Coincidentally, closes https://github.com/airbytehq/airbyte-internal-issues/issues/2469. After this PR, we'll trigger soft resets after raw table creation.

Extract some common code from the DefaultTyperDeduper and NoopTyperDeduperWithMigrations classes. This is just a refactor to simplify some future work.

It also fixes some tests in DefaultTyperDeduperTest. Previously some tests weren't actually doing anything, because they called clearInvocations() followed immediately by verifyNoMoreInteractions(), which is effectively just assertTrue(true).

The CDK build.gradle changes are just copied out of https://github.com/airbytehq/airbyte/pull/34637. Once that's merged, there should be no diff with master.

### CODE REVIEW

**Comment by edgao** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35376?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #35342** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35342?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #35376** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35376?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #35354** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35354?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/35376?utm_source=stack-comment).

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java [hunk 1]
```diff
@@ -108,36 +106,23 @@ public DefaultTyperDeduper(
...
-  public void prepareTables() throws Exception {
+  public void prepareFinalTables() throws Exception {
     if (overwriteStreamsWithTmpTable != null) {
       throw new IllegalStateException("Tables were already prepared.");
     }
     overwriteStreamsWithTmpTable = ConcurrentHashMap.newKeySet();
     LOGGER.info("Preparing tables");
 
-    // This is intentionally not done in parallel to avoid rate limits in some destinations.
-    prepareSchemas(parsedCatalog);
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java:131]: this call now happens in prepareSchemasAndRawTables

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java [hunk 2]
```diff
@@ -108,36 +106,23 @@ public DefaultTyperDeduper(
...
-
-    // TODO: Either the migrations run the soft reset and create v2 tables or the actual prepare tables.
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java:133]: this logic now happens in TyperDeduperUtil.executeRawTableMigrations

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/TyperDeduperUtil.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import com.google.common.collect.Streams
+import io.airbyte.cdk.integrations.util.ConnectorExceptionUtil.getResultsOrLogAndThrowFirst
+import io.airbyte.commons.concurrency.CompletableFutures
+import java.util.*
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionStage
+import java.util.concurrent.ExecutorService
+
+
+class TyperDeduperUtil {
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/TyperDeduperUtil.kt:12]: this class is partly taken from @jbfbell 's https://github.com/airbytehq/airbyte/pull/34637/files#diff-c9e7bf1074e965f63b135422dd2a7ac6ba196046b226a0ffe3a855a4eb82ed3b

it'll have some merge conflicts b/c I reordered the arguments in `prepareAllSchemas` but :shrug: just wanted to be consistent across the methods, and I generally prefer infra-ish params (executorservice, destinationhandler, etc.) to come before functional-ish params (parsedcatalog, streamconfig, etc.)

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/TyperDeduperUtil.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        v1V2Migrator: DestinationV1V2Migrator,
+        v2TableMigrator: V2TableMigrator,
+        parsedCatalog: ParsedCatalog
+    ) {
+      // TODO: Either the migrations run the soft reset and create v2 tables or the actual prepare tables.
+      // unify the logic
+      // with current state of raw tables & final tables. This is done first before gather initial state
+      // to avoid recreating
+      // final tables later again.
+      val runMigrationsResult =
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/TyperDeduperUtil.kt:29]: as noted above - this code originally lived in DefaultTyperDeduper, I just extracted it to a separate class.

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/test/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduperTest.java [hunk 5]
```diff
@@ -161,12 +165,14 @@ void existingEmptyTableMatchingSchema() throws Exception {
     initialStates.forEach(initialState -> {
       when(initialState.isFinalTablePresent()).thenReturn(true);
       when(initialState.isFinalTableEmpty()).thenReturn(true);
-      when(initialState.isSchemaMismatch()).thenReturn(true);
+      when(initialState.isSchemaMismatch()).thenReturn(false);
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/test/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduperTest.java:168]: this test previously didn't actually assert anything, and was secretly broken. This is the correct version.

(specifically: if there are no final tables, isSchemaMismatch is supposed to return false)

**Review by jbfbell — APPROVED** [review]

**Pull request merged**